### PR TITLE
crimson/seastore: fix segment_cleaner blocked_io_wake crash

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

File: src/crimson/os/seastore/segment_cleaner.h [hunk 1]
```diff
@@ -887,7 +926,17 @@ class SegmentCleaner : public SegmentProvider {
       [this] {
 	blocked_io_wake = seastar::promise<>();
 	return blocked_io_wake->get_future();
-      });
+      }
+    ).then([this, projected_usage] {
+      ceph_assert(!blocked_io_wake);
+      projected_used_bytes += projected_usage;
```
Comment (xxhdx1985126) [inline src/crimson/os/seastore/segment_cleaner.h:932]: It looks to me that the following scenario is still possible:

1. transaction A and B are both waiting on SegmentCleaner::blocked_io_wake
2. segment cleaner set SegmentCleaner::blocked_io_wake's value
3. transaction A is scheduled and set projected_used_bytes, which leads to another round of blocked io wait
4. transaction B is scheduled and the assert here is triggered

Am I right?

If so, I think this [condition variable](https://github.com/ceph/ceph/pull/41803/commits/3207f6413c15edc91e46c9cd154a84c93c78bbc8) should be able to help:-)

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.h:932]: Ah, sorry, this is protected by write_pipeline.reserve_projected_usage, right?

**Review by xxhdx1985126 — APPROVED** [review]
LGTM

File: src/crimson/os/seastore/segment_cleaner.h (hunk 1 above)
Comment (athanatos) [reply src/crimson/os/seastore/segment_cleaner.h:932]: Right.

**Pull request merged**